### ISSUE TITLE
Slow TAB completion in verdi commands

### ISSUE THREAD

**Issue by nmounet** [aiidateam/aiida_core#376]
All 'verdi' commands TAB completion have become very slow (1-2 seconds).

**Comment by nmounet** [comment]
To get an quantitative idea:

"time verdi shell  --help" gives

- in AiiDA v0.7.0: 
real	0m0.143s
user	0m0.080s
sys	0m0.052s

- in current develop:
real	0m0.918s
user	0m0.908s
sys	0m0.768s

**Comment by giovannipizzi** [comment]
See also discussion in #295 which is about the same problem

**Comment by nmounet** [comment]
Also, simply "time verdi help":
real	0m0.989s
user	0m0.956s
sys	0m0.692s

**Comment by giovannipizzi** [comment]
From profiling, this seems due to importing `pkg_resources`.
This is a problem seen also from many other people: just google e.g. for `import pkg_resources slow` or `load_entry_point slow`, and they have exactly the same problem (hundreds of ms to import, which is a huge problem for cmdline tools and even more for tab completion).

Anybody has suggestions on how to overcome the problem? @DropD ?

**Comment by giovannipizzi** [comment]
Another big bunch of time (~0.4 sec on my computer, about 1/3 of the time to run 'verdi user list') is due to the import of aiida.optional. It looks to me that it was introduced by @muhrin in 246d0a63a55a, and it is to know if kombu is installed, and in this case add it in settings.py for Django.

@muhrin is this still needed? If so, can we just (directly in the `settings.py`) do a try/except of `import kombu` that should be much faster?

**Comment by giovannipizzi** [comment]
For comparison:
```
$ time python -c 'import kombu'

real	0m0.022s
user	0m0.012s
sys	0m0.007s

$ time python -c 'import aiida.optional'

real	0m0.336s
user	0m0.255s
sys	0m0.074s
```

**Comment by giovannipizzi** [comment]
By the way, if you want to do some more serious profiling, `pip install snakeviz` as it is a very nice graphical tool

**Comment by szoupanos** [comment]
I will also try to have a look at this. I agree that the tab completion is slow and I think that it becomes worse when the database gets bigger (I don't know how these 2 facts could be related - I have to check).

**Comment by DropD** [comment]
We can easily get rid of the pkg_resource import. Actually installing aiida using `pip wheel` ("built" package) generates a script for the verdi entry point that does not install pkg_resources. `time verdi` returns 0.15 for me that way.

wheel installs are not editable (not in the same way at least), so this is not very useful for everyone.
setuptools may switch to using the same system for source installs in the future, but that only helps people with an up to date setuptools.

The reason why we don't just want to install verdi as a script instead of an entry point is that the entry point ensures that verdi always runs with the right environment.

The choice we have seems to be:
A) running verdi without activating the virtual environment works correctly
  * only fast if installed as wheel (non-editable)
  * quite slow otherwise
B) if verdi is run without activating the environment, undefined behaviour happens (confusing the user)
  * if run correctly it will be 0.3-0.4 s faster in both cases
  * it might still be even faster when installed as wheel
C) we emulate the script-generation that `pip wheel` does ourselves
  * best of both worlds
  * potentially hard to maintain
  * potentially hard to implement / verify

I would prefer A) over B) long term, because the long term goal should be that users will never ever touch the source code of their production aiida instance. However that will only be the case when AiiDA ships with the new plugin system. 
Therefore it might be best to investigate C) first, and if it turns out to be hard, go with B) until we have a stable version that can be wheel installed by users.

**Comment by DropD** [comment]
route C) has already been implemented, PR #425 enables it for aiida. It makes verdi start nearly as fast as using a wheel install.

**Comment by nmounet** [comment]
Ricoh's improvementsare great and help a lot. Yet, I'm not sure we can close the issue, as the tab completion is still noticeable for me even after rerunning `pip install` (`verdi --help` takes 0.4s)...

**Comment by DropD** [comment]
It must be made a strict rule that command modules may never import from aiida on module level, only in functions.

verdi was again recently slowed down because aiida/cmdline/commands/work.py was importing aiida.utils.ascii_vis

Simply importing that module timed at 0.3 seconds for me and it is only used inside a function. New pull request upcoming.

**Comment by nmounet** [comment]
It's slow again...

> time verdi --help
```
real	0m0.648s
user	0m0.616s
sys	0m0.024s

```

python -m profile --sort=cum ~/aiidapy/bin/verdi|more

**Comment by DropD** [comment]
from the profiling log it looks like the verdi instance in question was installed before we started using fastentrypoints.

Are you sure you have reinstalled aiida-core in the virtualenv in question, from a branch that has fastentrypoints.py?

If yes, somehow an import pkg_ressources must have snuck into cmdline somewhere

**Comment by nmounet** [comment]
You are right. For some reason I forgot to re-install AiiDA within this particular virtual env. Apologies.
Issue was then already solved, I close it (again...)

**Comment by szoupanos** [comment]
At my desktop, I see that the autocompletion doesn't work at all.

I ensured by executing again the

> $ pip install --user -U setuptools pip wheel virtualenv

that the venv was installed correctly.

I installed (again) aiida as it is shown at http://aiida-core.readthedocs.io/en/latest/installation/index.html
> (aiidapy) $ pip install -e aiida_core

And I also executed a 

> verdi setup

 to reconfigure a specific profile (I don't know if this changes anything).

So after doing the above, I still don't have autocompletion when writing half of the command and I press tab. On my laptop, on the other hand, autocompletion works and it is actually fast.

Any ideas?

Thanks a lot!
Spyros

**Comment by szoupanos** [comment]
I created a new issue (#498) for my last comment since the problem is not that the autocompletion is slow but that it does't work at all.

**Comment by sphuber** [comment]
Due to incorrect imports `verdi` is once again slow:

```
(aiida_hp) sphuber@theos:~/code/aiida/env/hp/aiida-core$ time verdi --help
verdi: '--help' is not a valid command. See 'verdi help' for more help.

Did you mean this?
     help

real	0m0.602s
user	0m0.496s
sys	0m0.096s
```

**Comment by sphuber** [comment]
Another occurrence of this problem was report in #1793

**Comment by giovannipizzi** [comment]
`time COMP_WORDS="verdi devel tests db.cmdline" COMP_CWORD=3 _VERDI_COMPLETE=complete-bash python `which verdi` is slow again...:
```
real	0m1.487s
user	0m0.863s
sys	0m0.602s
```

**Comment by giovannipizzi** [comment]
A comment: maybe a (weird?) way to prevent this from happening is to define a decorator `@without_dbenv` that checks (before running the function and after) that the dbenv is NOT loaded and raises in case it is. This will prevent forever that that function accidentally loads dbenv.

loading dbenv is not the only source of possible problems, but it's definitely one of the major ones.

**Comment by sphuber** [comment]
Closed once more in #2315
